### PR TITLE
re-try git clone when copr-dist-git is temporarily down

### LINKED ISSUE #2394: We should re-try git clone when copr-dist-git is temporarily down
https://download.copr.fedorainfracloud.org/results/iucar/cran/fedora-37-x86_64/05072406-R-CRAN-sparsegl/builder-live.log.gz
```
Running: git clone https://copr-dist-git.fedorainfracloud.org/git/iucar/cran/R-CRAN-sparsegl /var/lib/copr-rpmbuild/workspace/workdir-_250bprx/R-CRAN-sparsegl --depth 500 --no-single-branch --recursive

cmd: ['git', 'clone', 'https://copr-dist-git.fedorainfracloud.org/git/iucar/cran/R-CRAN-sparsegl', '/var/lib/copr-rpmbuild/workspace/workdir-_250bprx/R-CRAN-sparsegl', '--depth', '500', '--no-single-branch', '--recursive']
cwd: .
rc: 128
stdout: 
stderr: Cloning into '/var/lib/copr-rpmbuild/workspace/workdir-_250bprx/R-CRAN-sparsegl'...
fatal: unable to access 'https://copr-dist-git.fedorainfracloud.org/git/iucar/cran/R-CRAN-sparsegl/': Failed to connect to copr-dist-git.fedorainfracloud.org port 443 after 1 ms: Connection refused

Cloning into '/var/lib/copr-rpmbuild/workspace/workdir-_250bprx/R-CRAN-sparsegl'...
fatal: unable to access 'https://copr-dist-git.fedorainfracloud.org/git/iucar/cran/R-CRAN-sparsegl/': Failed to connect to copr-dist-git.fedorainfracloud.org port 443 after 1 ms: Connection refused
Running: git clone https://copr-dist-git.fedorainfracloud.org/git/iucar/cran/R-CRAN-sparsegl /var/lib/copr-rpmbuild/workspace/workdir-_250bprx/R-CRAN-sparsegl

cmd: ['git', 'clone', 'https://copr-dist-git.fedorainfracloud.org/git/iucar/cran/R-CRAN-sparsegl', '/var/lib/copr-rpmbuild/workspace/workdir-_250bprx/R-CRAN-sparsegl']
cwd: .
rc: 128
stdout: 
stderr: Cloning into '/var/lib/copr-rpmbuild/workspace/workdir-_250bprx/R-CRAN-sparsegl'...
fatal: unable to access 'https://copr-dist-git.fedorainfracloud.org/git/iucar/cran/R-CRAN-sparsegl/': Failed to connect to copr-dist-git.fedorainfracloud.org port 443 after 2 ms: Connection refused


Traceback (most recent call last):
  File "/usr/lib/python3.11/site-packages/copr_rpmbuild/helpers.py", line 286, in git_clone_and_checkout
    run_cmd(clone_cmd)
  File "/usr/lib/python3.11/site-packages/copr_rpmbuild/helpers.py", line 74, in run_cmd
    raise RuntimeError(result.stderr)
RuntimeError: Cloning into '/var/lib/copr-rpmbuild/workspace/workdir-_250bprx/R-CRAN-sparsegl'...
fatal: unable to access 'https://copr-dist-git.fedorainfracloud.org/git/iucar/cran/R-CRAN-sparsegl/': Failed to connect to copr-dist-git.fedorainfracloud.org port 443 after 1 ms: Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/bin/copr-rpmbuild", line 259, in build_rpm
    distgit.produce_sources()
  File "/usr/lib/python3.11/site-packages/copr_rpmbuild/providers/distgit.py", line 34, in produce_sources
    helpers.git_clone_and_checkout(self.clone_url, self.committish,
  File "/usr/lib/python3.11/site-packages/copr_rpmbuild/helpers.py", line 291, in git_clone_and_checkout
    run_cmd(['git', 'clone', url, repo_path])
  File "/usr/lib/python3.11/site-packages/copr_rpmbuild/helpers.py", line 74, in run_cmd
    raise RuntimeError(result.stderr)
RuntimeError: Cloning into '/var/lib/copr-rpmbuild/workspace/workdir-_250bprx/R-CRAN-sparsegl'...
fatal: unable to access 'https://copr-dist-git.fedorainfracloud.org/git/iucar/cran/R-CRAN-sparsegl/': Failed to connect to copr-dist-git.fedorainfracloud.org port 443 after 2 ms: Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/bin/copr-rpmbuild", line 134, in main
    action(args, config)
  File "/usr/bin/copr-rpmbuild", line 266, in build_rpm
    builder.archive_configs()
    ^^^^^^^
UnboundLocalError: cannot access local variable 'builder' where it is not associated with a value

```

### PR DESCRIPTION
**Pull request by nikromen**
how copr_rpmbiuld is run? If I throw there a sleep (as I does) do I stop building only for one VM (and one user) or for everybody?

Fix #2394

### CODE REVIEW

**Comment by xsuchy** [comment]
OMG this is so cool. It will be useful for others... Can we put it in separate module and put on pypi. Oh wait... https://pypi.org/search/?q=retry Can we use one of these?

**Comment by nikromen** [comment]
Oh, I wouldn't think of looking for a package for solution that is 15 lines long :D

I tried to contact maintainer of https://src.fedoraproject.org/rpms/python-retry to consider whether he wants to rebase the package to its maintained fork https://github.com/eSAMTrade/retry (original upstream is unmaintained for 6 years). In case of no response I will package it by myself.

I am closing this PR since it is not needed in this way.

**Comment by nikromen** [comment]
Or wait... I can rebase this PR (with using the retry package)

sorry for the confusion :D

**Comment by praiskup** [comment]
From mtg: There's also python3-reretry in Fedora repos.

File: rpmbuild/copr-rpmbuild.spec [hunk 1]
```diff
@@ -47,6 +47,12 @@ BuildRequires: %{python_pfx}-munch
 BuildRequires: %{python}-requests
 BuildRequires: %{python_pfx}-jinja2
 BuildRequires: %{python_pfx}-simplejson
+%if 0%{?fedora}
+BuildRequires: %{python}-retry
+%endif
+%if 0%{?rhel}
+BuildRequires: %{python}-retrying
```
Comment (praiskup) [inline rpmbuild/copr-rpmbuild.spec:54]: Can we have a dummy test-case that the lib works?

Comment (nikromen) [reply rpmbuild/copr-rpmbuild.spec]: The lib doesn't work (our irc chat). But I found the exact same package in epel https://bodhi.fedoraproject.org/updates/FEDORA-EPEL-2023-46680e2ac7

**Comment by nikromen** [comment]
hm, but the retry package is missing for epel8

**Comment by praiskup** [comment]
Have you considered using https://github.com/litl/backoff ?   I somewhat like it's interface even more.

**Comment by nikromen** [comment]
As I see it, the backoff package https://src.fedoraproject.org/rpms/python-backoff doesn't have EPEL9 and has a rather outdated version (1.10 vs 2.2.1), which means going through the same cycle again which I am not happy about (I would rather package the retry in copr for epel8)... but since the maintainer is in RH I'll write to him.

**Comment by praiskup** [comment]
Just request a rebuild in epel9?  Do we need the new version?

**Comment by praiskup** [comment]
I mean, of course, I have no problem with either of them - and it shouldn't be a problem to build either of them in epel8+.  Shouldn't be a long shot for us.

**Comment by praiskup** [comment]
+1

**Pull request merged**